File: backend/src/main/java/com/capstone/dsa/MergeSort.java
package com.capstone.dsa;

import java.util.Comparator;
import java.util.List;
import java.util.ArrayList;

/**
 * Generic Merge Sort implementation for multi-field leaderboard ranking.
 *
 * <pre>
 * WHY MERGE SORT?
 *   - Guaranteed O(N log N) — no worst-case degradation (unlike QuickSort)
 *   - Stable sort — teams with equal scores maintain their original order
 *   - Well-suited for Java generics with a custom Comparator
 *
 * APPLIED TO:
 *   1. Leaderboard: sort teams by (totalScore DESC, teamName ASC)
 *   2. Mentorship requests: sort by (matchScore DESC, requestedAt ASC)
 *   3. Project ideas: sort by (upvoteCount DESC, createdAt DESC)
 * </pre>
 *
 * Time:  O(N log N) — always
 * Space: O(N)       — temporary arrays in merge step
 */
public class MergeSort {

    // Private constructor — utility class, not instantiable
    private MergeSort() {}

    /**
     * Sort a list using the provided Comparator.
     * Returns a new sorted list; original list is not mutated.
     *
     * @param list       Input list
     * @param comparator Comparison function defining sort order
     * @param <T>        Element type
     * @return New sorted list
     */
    public static <T> List<T> sort(List<T> list, Comparator<T> comparator) {
        if (list == null || list.size() <= 1) {
            return new ArrayList<>(list == null ? List.of() : list);
        }
        return mergeSort(new ArrayList<>(list), comparator);
    }

    // ── Recursive helpers ─────────────────────────────────────────────────────

    private static <T> List<T> mergeSort(List<T> list, Comparator<T> cmp) {
        if (list.size() <= 1) return list;                  // Base case

        int mid = list.size() / 2;
        List<T> left  = mergeSort(new ArrayList<>(list.subList(0, mid)),        cmp);
        List<T> right = mergeSort(new ArrayList<>(list.subList(mid, list.size())), cmp);

        return merge(left, right, cmp);
    }

    /**
     * Core merge step — combines two sorted halves into one sorted list.
     * Time: O(N) where N = left.size() + right.size()
     */
    private static <T> List<T> merge(List<T> left, List<T> right, Comparator<T> cmp) {
        List<T> result = new ArrayList<>(left.size() + right.size());
        int i = 0, j = 0;

        while (i < left.size() && j < right.size()) {
            // cmp.compare returns ≤ 0 when left[i] should come before right[j]
            if (cmp.compare(left.get(i), right.get(j)) <= 0) {
                result.add(left.get(i++));
            } else {
                result.add(right.get(j++));
            }
        }

        // Append any remaining elements (at most one loop will execute)
        while (i < left.size())  result.add(left.get(i++));
        while (j < right.size()) result.add(right.get(j++));

        return result;
    }

    // ── Pre-built comparators for the application domain ─────────────────────

    /**
     * Leaderboard comparator:
     *   1. totalScore DESC (higher = better rank)
     *   2. teamName    ASC (alphabetical tiebreaker)
     */
    public static <T> Comparator<T> leaderboardComparator(
            java.util.function.ToIntFunction<T>    scoreExtractor,
            java.util.function.Function<T, String> nameExtractor
    ) {
        return Comparator
                .comparingInt(scoreExtractor).reversed()        // DESC score
                .thenComparing(nameExtractor);                  // ASC name
    }
}

File: matching-service/dsa/merge_sort.py
"""
dsa/merge_sort.py
───────────────────────────────────────────────────────────────────────────────
Generic Multi-Key Merge Sort for the Leaderboard ranking system.

WHY CUSTOM MERGE SORT?
  • Python's built-in sort uses Timsort (also O(N log N)) — we implement our
    own to demonstrate the algorithm explicitly for the mini-project.
  • Our version supports multiple sort keys with per-key sort direction,
    which is needed for leaderboard tiebreaking:
        Primary   → total_score (DESC)
        Secondary → milestone_completion_pct (DESC)
        Tertiary  → team_name (ASC alphabetical)

COMPLEXITY:
  • Time:  O(N log N) — always (no worst case like quicksort)
  • Space: O(N)       — merge step requires temporary arrays

The merge sort is also used to rank idea similarity results (descending
by TF-IDF cosine score) in the idea_similarity router.
"""

from typing import Any, Callable, List, TypeVar

T = TypeVar("T")


def merge_sort(
    arr: List[T],
    key: Callable[[T], Any] = lambda x: x,
    reverse: bool = False,
) -> List[T]:
    """
    Recursive merge sort with a custom key function and sort direction.

    Args:
        arr:     Input list (not modified in place — returns new list)
        key:     Key extractor function, default identity
        reverse: True = descending, False = ascending

    Returns:
        New sorted list
    """
    if len(arr) <= 1:
        return list(arr)  # Base case

    mid   = len(arr) // 2
    left  = merge_sort(arr[:mid],  key=key, reverse=reverse)
    right = merge_sort(arr[mid:],  key=key, reverse=reverse)

    return _merge(left, right, key, reverse)


def _merge(left: List[T], right: List[T], key: Callable, reverse: bool) -> List[T]:
    """
    Merge two sorted lists into one sorted list.
    This is the core O(N) step of merge sort.
    """
    result: List[T] = []
    i = j = 0

    while i < len(left) and j < len(right):
        lk = key(left[i])
        rk = key(right[j])

        # Take from left if:
        #   ascending  → left_key ≤ right_key
        #   descending → left_key ≥ right_key
        if (not reverse and lk <= rk) or (reverse and lk >= rk):
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # Append remaining elements
    result.extend(left[i:])
    result.extend(right[j:])
    return result


def multi_key_sort(arr: List[dict], sort_spec: List[dict]) -> List[dict]:
    """
    Sort a list of dicts by multiple keys with independent directions.

    Args:
        arr:       List of dicts to sort
        sort_spec: List of {"key": str, "reverse": bool} dicts,
                   applied left-to-right (first = primary sort key)

    Example:
        multi_key_sort(teams, [
            {"key": "total_score",             "reverse": True},   # DESC
            {"key": "milestone_completion_pct","reverse": True},   # DESC
            {"key": "team_name",               "reverse": False},  # ASC
        ])
    """
    # Apply specs from LAST to FIRST (stable sort maintains order)
    result = list(arr)
    for spec in reversed(sort_spec):
        field   = spec["key"]
        reverse = spec.get("reverse", False)
        result  = merge_sort(result, key=lambda x, f=field: x.get(f, 0), reverse=reverse)
    return result


def rank_candidates(candidates: List[dict], score_field: str = "match_score") -> List[dict]:
    """
    Sort matching candidates by score (descending) using merge sort.
    Adds a "rank" field (1-indexed) to each candidate.

    Used by:
      - teammate_match router  (sort by complementarity score)
      - mentor_match router    (sort by adjusted_score)
      - idea_similarity router (sort by cosine similarity)
    """
    sorted_candidates = merge_sort(
        candidates,
        key=lambda c: c.get(score_field, 0.0),
        reverse=True,   # Highest score first
    )
    for idx, candidate in enumerate(sorted_candidates):
        candidate["rank"] = idx + 1
    return sorted_candidates
